### PR TITLE
Run interactive session as a non-root user

### LINKED ISSUE #3051: Run cl run --interactive as a non-root user
**Describe the bug**
Run `cl run --interactive` as a non-root user to match the same behavior as CodaLab worker running jobs.

### PR DESCRIPTION
**Pull request by teetone**
### Reasons for making this change

- Run `cl run --interactive` as a non-root user
- Mount a bash history file and read from it directly after the end of an interactive session

### Related issues

Resolves #3051

### Checklist

* [x] I've added a screenshot of the changes, if this is a frontend change
* [x] I've added and/or updated tests, if this is a backend change
* [x] I've run the [pre-commit.sh](https://github.com/codalab/codalab-worksheets/blob/master/pre-commit.sh) script
* [x] I've updated [docs](https://github.com/codalab/codalab-worksheets/tree/master/docs), if needed

### CODE REVIEW

File: codalab/lib/interactive_session.py [hunk 1]
```diff
@@ -136,14 +142,23 @@ def get_docker_path(sub_path):
...
         name = self._get_container_name()
-        command = ['docker run', '-it', f'--name {name}', f'-w {os.path.sep}{self._session_uuid}']
+        # Start a container as a non-root user. Root (id = 0) is the default user within a container.
+        # When passing a numeric ID, the user does not have to exist in the container.
+        command = [
+            'docker run',
+            '-it',
+            f'--name {name}',
+            f'-w {os.path.sep}{self._session_uuid}',
+            '-u 1',
```
Comment (epicfaace) [inline codalab/lib/interactive_session.py:153]: I'm not too familiar with UIDs, but why pick 1 (daemon) as the UID?

[This page](https://refspecs.linuxbase.org/LSB_3.1.1/LSB-Core-generic/LSB-Core-generic/usernames.html#FTN.AEN26413) says: "The daemon User ID/Group ID was used as an unprivileged User ID/Group ID for daemons to execute under in order to limit their access to the system. Generally daemons should now run under individual User ID/Group IDs in order to further partition daemons from one another."

File: codalab/lib/interactive_session.py [hunk 2]
```diff
@@ -175,10 +190,9 @@ def _get_container_name(self):
     def _construct_final_command(self):
         try:
             candidate_commands = self._get_bash_history()
-        except Exception:
+        except Exception as e:
             print(
-                'The history of bash commands could not be retrieved at path: %s'
-                % InteractiveSession._BASH_HISTORY_CONTAINER_PATH,
+                'The history of bash commands could not be retrieved at path: {}'.format(e),
```
Comment (epicfaace) [inline codalab/lib/interactive_session.py]: Did you mean to include `InteractiveSession._BASH_HISTORY_CONTAINER_PATH` in this error message as well?

File: tests/unit/lib/interactive_session_test.py [hunk 3]
```diff
@@ -28,11 +29,14 @@ def test_get_docker_run_command_with_subpaths(self):
...
         )
+        session._host_bash_history_path = ".bash_history"
         expected_regex = (
-            'docker run -it --name interactive-session-0x[a-z0-9]{32} -w \/0x[a-z0-9]{32} -v '
+            'docker run -it --name interactive-session-0x[a-z0-9]{32} -w \/0x[a-z0-9]{32} -u 1 -v '
             '[\s\S]{0,100}local\/path1/sub/path1:\/0x[a-z0-9]{32}\/key:ro -v [\s\S]{0,100}local\/path2/sub/path2'
-            ':\/0x[a-z0-9]{32}\/key2:ro some-docker-image bash'
+            ':\/0x[a-z0-9]{32}\/key2:ro -v \.bash_history:\/usr\/sbin\/\.bash_history:rw some-docker-image bash'
         )
+        print(expected_regex)
```
Comment (epicfaace) [inline tests/unit/lib/interactive_session_test.py]: remove print statements

File: codalab/lib/interactive_session.py [hunk 4]
```diff
@@ -95,7 +99,9 @@ def start(self):
             print('CodaLab instance:', self._manager.current_client().address, file=self._stdout)
             print('Container name:', self._get_container_name(), file=self._stdout)
             print('Container Docker image:', self._docker_image, file=self._stdout)
-            print('You can find local bundle contents at:', self._bundle_path, file=self._stdout)
+            print(
+                'You can find local bundle contents at:', self._host_bundle_path, file=self._stdout
```
Comment (epicfaace) [inline codalab/lib/interactive_session.py]:
```suggestion
                'You can find local bundle contents at: ', self._host_bundle_path, file=self._stdout
```

File: codalab/lib/interactive_session.py [hunk 5]
```diff
@@ -80,10 +79,15 @@ def __init__(
...
     def start(self):
-        self._bundle_path = os.path.join(
+        self._host_bundle_path = os.path.join(
             self._manager.codalab_home, 'local_bundles', self._session_uuid
         )
-        os.makedirs(self._bundle_path)
+        os.makedirs(self._host_bundle_path)
+
+        # Create a blank file which will be used as the bash history file that will later be
+        # mounted and populated during the interactive session.
```
Comment (epicfaace) [inline codalab/lib/interactive_session.py:89]: What's the purpose of mounting `.bash_history`?

Comment (teetone) [reply codalab/lib/interactive_session.py:89]: We mount `.bash_history` so we have access to the file on the host after the container exits.

File: codalab/lib/interactive_session.py (hunk 1 above)
Comment (teetone) [reply codalab/lib/interactive_session.py:153]: 0 is for root and any non-zero number is for a non-root user. You can specify a number without having to register a user.

**Review by yiboliu — APPROVED** [review]

**Pull request merged**